NetSerializer/Primitives.cs

```diff
@@ -318,7 +324,122 @@ public static void ReadPrimitive(Stream stream, out decimal value)
 			value = new Decimal(arr);
 		}
 
-#if NO_UNSAFE
+#if NETCOREAPP
+		public static void WritePrimitive(Stream stream, string value)
+		{
+			if (value == null)
+			{
+				WritePrimitive(stream, (uint)0);
+				return;
+			}
+
+			if (value.Length == 0)
+			{
+				WritePrimitive(stream, (uint)1);
+				return;
+			}
+
+			Span<byte> buf = stackalloc byte[StringByteBufferLength];
+
+			var totalChars = value.Length;
+			var totalBytes = Encoding.UTF8.GetByteCount(value);
+
+			WritePrimitive(stream, (uint)totalBytes + 1);
+			WritePrimitive(stream, (uint)totalChars);
+
+			var totalRead = 0;
+			ReadOnlySpan<char> span = value;
+			for (;;)
+			{
+				var finalChunk = totalRead + totalChars >= totalChars;
+				Utf8.FromUtf16(span, buf, out var read, out var wrote, isFinalBlock: finalChunk);
+				stream.Write(buf.Slice(0, wrote));
+				totalRead += read;
+				if (read >= totalChars)
+				{
+					break;
+				}
+
+				span = span[read..];
+				totalChars -= read;
+			}
+		}
+
+		private static readonly SpanAction<char, (int, Stream)> _stringSpanRead = StringSpanRead;
+
+		public static void ReadPrimitive(Stream stream, out string value)
+		{
+			ReadPrimitive(stream, out uint totalBytes);
+
+			if (totalBytes == 0)
+			{
+				value = null;
+				return;
+			}
+
+			if (totalBytes == 1)
+			{
+				value = string.Empty;
+				return;
+			}
+
+			totalBytes -= 1;
+
+			ReadPrimitive(stream, out uint totalChars);
+
+			value = string.Create((int) totalChars, ((int) totalBytes, stream), _stringSpanRead);
+		}
+
+		private static void StringSpanRead(Span<char> span, (int totalBytes, Stream stream) tuple)
+		{
+			Span<byte> buf = stackalloc byte[StringByteBufferLength];
+
+			// ReSharper disable VariableHidesOuterVariable
+			var (totalBytes, stream) = tuple;
+			// ReSharper restore VariableHidesOuterVariable
```

Comment:
What's the outer variable this is hiding?

Reply:
Oh this might've actually been a holdover from when I still had this method as a closure instead.